src/Search/SearchResult.php

```diff
@@ -84,16 +94,53 @@ public function __construct(array $body)
      *
      * @return SearchResult
      */
-    public function filter(callable $callback): self
+    public function applyOptions($options): self
+    {
+        if (\array_key_exists('removeZeroFacets', $options) && true === $options['removeZeroFacets']) {
+            $this->removeZeroFacets();
+        }
+        if (\array_key_exists('transformHits', $options) && \is_callable($options['transformHits'])) {
+            $this->transformHits($options['transformHits']);
+        }
+        if (\array_key_exists('transformFacetsDistribution', $options) && \is_callable($options['transformFacetsDistribution'])) {
+            $this->transformFacetsDistribution($options['transformFacetsDistribution']);
+        }
+
+        return $this;
+    }
+
+    public function transformHits(callable $callback): self
     {
-        $results = array_filter($this->hits, $callback, ARRAY_FILTER_USE_BOTH);
+        $this->hits = $callback($this->hits);
+        $this->hitsCount = \count($this->hits);
 
-        $this->hits = $results;
-        $this->nbHits = \count($results);
+        return $this;
+    }
+
+    public function transformFacetsDistribution(callable $callback): self
+    {
+        $this->facetsDistribution = $callback($this->facetsDistribution);
 
         return $this;
     }
 
+    public function removeZeroFacets(): self
+    {
+        $filterAllFacets = function (array $facets) {
+            $filterOneFacet = function (array $facet) {
+                return array_filter(
+                    $facet,
+                    function ($v, $k) { return 0 !== $v; },
+                    ARRAY_FILTER_USE_BOTH
+                );
+            };
+
+            return array_map($filterOneFacet, $facets);
+        };
+
+        return $this->transformFacetsDistribution($filterAllFacets);
```

Comment:
Just curiosity
Would this be a possible or better alternative?

```php
$this->facetsDistribution = $filterAllFacets($facetsDistribution);

return $this
```

or even this: 

```php

 public function removeZeroFacets(): self
    {
        $filterAllFacets = function (array $facets) {
            $filterOneFacet = function (array $facet) {
                return array_filter(
                    $facet,
                    function ($v, $k) { return 0 !== $v; },
                    ARRAY_FILTER_USE_BOTH
                );
            };

            $this->transformFacetsDistribution = array_map($filterOneFacet, $facets);
        };

        $filterAllFacets($this->facetsDistribution);

        return $this;
}
```

In both cases `transformFacetsDistribution` wouldn't be useful. But If you want to use the method you created I understand. But I think in either cases you can return `$this` no ? instead of 
```php
return $this->transformFacetsDistribution($filterAllFacets);
```
which result is not used

Reply:
I would rather use the `transformFacetsDistribution` function.
Indeed, if at some point, there is another field to update when transforming the `facetsDistribution` (as it is for `transformHits` where we update the `hits` AND the `hitsCount`), I would rather all be done in the same function to avoid duplication, so to avoid mistakes.

Reply:
And about changing this to  `  return $this->transformFacetsDistribution($filterAllFacets);`

```php
    $this->transformFacetsDistribution($filterAllFacets);

    return $this;
```

which would avoid making this return the facetDistribution and thus changing the value of searchResponse.
```php
$searchResponse = $this-removeZeroFacets() 
```


This would avoid having to do this:

Reply:
If I'm not wrong, maybe I miss-understand something, but here, your alternative and my original suggestion are exactly the same